model/jpa/src/main/java/org/keycloak/connections/jpa/PersistenceExceptionConverter.java

```diff
@@ -95,6 +98,8 @@ public static ModelException convert(Throwable t) {
             throw new ModelDuplicateException(t.getCause());
         } else if (throwModelDuplicateEx.test(t)) {
             throw new ModelDuplicateException(t);
+        } else if (t instanceof HibernateException || t instanceof OptimisticLockException) {
+            throw new RuntimeException("Database operation failed", t);
```

Comment:
I had a look at the existing exceptions, and `ModelIllegalStateException` might be a better fit than a `RuntimeException`. It might even fix the other problem you're seeing with the failing test (I didn't try it though).

Catching the `HibernateException` here seems to be a bit broad here, I'd like to restrict it to the `OptimisticLockException`. 

Looking at the original issue that the message of a ModelException is passed to the caller without modification, it seems that it is a bad idea to pass on the exception to the constructor without providing a message. So I'd suggest to remove that constructor and only use the one which explicitly provides a message. WDYT?

Reply:
> I had a look at the existing exceptions, and ModelIllegalStateException might be a better fit than a RuntimeException. 


`ModelIllegalStateException` seems better to me as well.
I was previouly thinking this (from issue description) "_There is existing exception ModelIllegalStateException that could in theory be used, but it is subclass of ModelException and therefore as of today it is already mapped to 400. Changing that would mean we should add specific catch for ModelIllegalStateException to all relevant places to return 500 which does not seem right._" but maybe this is the right approach.

> It might even fix the other problem you're seeing with the failing test (I didn't try it though).
  
I needed to add catches to places that already catched the parent class `ModelException` and mapped it to `400`.

Now I have explicit catch `ModelIllegalStateException` and return `500`

> Catching the HibernateException here seems to be a bit broad here, I'd like to restrict it to the OptimisticLockException.

Done!


> Looking at the original issue that the message of a ModelException is passed to the caller without modification, it seems that it is a bad idea to pass on the exception to the constructor without providing a message. So I'd suggest to remove that constructor and only use the one which explicitly provides a message. WDYT? 

I removed the constructor and changed all accordingly.